cstore_writer.c

```diff
@@ -239,9 +326,10 @@ CStoreWriteRow(TableWriteState *writeState, Datum *columnValues, bool *columnNul
 			Oid columnCollation = attributeForm->attcollation;
 
 			blockData->existsArray[blockRowIndex] = true;
-			blockData->valueArray[blockRowIndex] = DatumCopy(columnValues[columnIndex],
-															 columnTypeByValue,
-															 columnTypeLength);
+
+			SerializeSingleDatum(blockData->uncompressedDataBuffer,
+								 columnValues[columnIndex], columnTypeByValue,
+								 columnTypeLength, attributeForm->attalign);
```

Comment:
although buffer enlargements seemed to be done incrementally, it does allocate in chunks i.e. twice as the existing memory amount. It preallocates 1024 bytes in the beginning, and doubles the memory as needed during enlarge operation.